File: src/main/java/me/alegian/thavma/impl/common/block/entity/WorkbenchBE.java
package me.alegian.thavma.impl.common.block.entity;

import me.alegian.thavma.impl.init.registries.deferred.T7BlockEntities;
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import software.bernie.geckolib.animatable.GeoBlockEntity;
import software.bernie.geckolib.animatable.SingletonGeoAnimatable;
import software.bernie.geckolib.animatable.instance.AnimatableInstanceCache;
import software.bernie.geckolib.animation.AnimatableManager;
import software.bernie.geckolib.animation.AnimationController;
import software.bernie.geckolib.animation.PlayState;
import software.bernie.geckolib.animation.RawAnimation;
import software.bernie.geckolib.util.GeckoLibUtil;

public class WorkbenchBE extends BlockEntity implements GeoBlockEntity {
  private final AnimatableInstanceCache cache = GeckoLibUtil.createInstanceCache(this);
  private final AnimationController<WorkbenchBE> ANIM_CONTROLLER = new AnimationController<>(this, "cycle", 20, state -> PlayState.CONTINUE)
      .triggerableAnim("closed", RawAnimation.begin().thenLoop("closed"))
      .triggerableAnim("open", RawAnimation.begin().thenLoop("open"))
      .triggerableAnim("rotating", RawAnimation.begin().thenLoop("rotating"));

  /**
   * Dummy constructor used for rendering Item form
   */
  public WorkbenchBE() {
    this(new BlockPos(0, 0, 0), T7Blocks.ARCANE_WORKBENCH.get().defaultBlockState());
  }

  public WorkbenchBE(BlockPos pos, BlockState blockState) {
    super(T7BlockEntities.INSTANCE.getWORKBENCH().get(), pos, blockState);
    SingletonGeoAnimatable.registerSyncedAnimatable(this);
  }

  @Override
  public void registerControllers(AnimatableManager.ControllerRegistrar controllers) {
    controllers.add(this.ANIM_CONTROLLER);
  }

  @Override
  public AnimatableInstanceCache getAnimatableInstanceCache() {
    return this.cache;
  }
}

File: src/main/java/me/alegian/thavma/impl/common/block/entity/WorkbenchBE.kt
package me.alegian.thavma.impl.common.block.entity

import me.alegian.thavma.impl.init.registries.deferred.T7BlockEntities.WORKBENCH
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ARCANE_WORKBENCH
import net.minecraft.core.BlockPos
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import software.bernie.geckolib.animatable.GeoBlockEntity
import software.bernie.geckolib.animatable.SingletonGeoAnimatable
import software.bernie.geckolib.animatable.instance.AnimatableInstanceCache
import software.bernie.geckolib.animation.AnimatableManager.ControllerRegistrar
import software.bernie.geckolib.animation.AnimationController
import software.bernie.geckolib.animation.RawAnimation
import software.bernie.geckolib.util.GeckoLibUtil

/**
 * Default values used for rendering Item form
 */
class WorkbenchBE(
    pos: BlockPos = BlockPos(0, 0, 0),
    blockState: BlockState = ARCANE_WORKBENCH.get().defaultBlockState()
) : BlockEntity(WORKBENCH.get(), pos, blockState), GeoBlockEntity {
    var playerInRange: Player? = null
    private val cache = GeckoLibUtil.createInstanceCache(this)
    val CLOSED_ANIM = RawAnimation.begin().thenLoop("closed")
    val OPEN_ANIM = RawAnimation.begin().thenLoop("rotating")
    private val ANIM_CONTROLLER = AnimationController(
        this, "main", 20
    ) { state ->
        var animation: RawAnimation? = state.controller.currentRawAnimation
        if (animation == null) {
            animation = if(isPlayerNearby()) OPEN_ANIM else CLOSED_ANIM
        }
        state.setAndContinue(animation)
    }
        .triggerableAnim("closed", CLOSED_ANIM)
        .triggerableAnim("rotating", OPEN_ANIM)

    init {
        SingletonGeoAnimatable.registerSyncedAnimatable(this)
    }

    override fun registerControllers(controllers: ControllerRegistrar) {
        controllers.add(this.ANIM_CONTROLLER)
    }

    override fun getAnimatableInstanceCache(): AnimatableInstanceCache {
        return this.cache
    }

    private fun isPlayerNearby(): Boolean {
        level.let {
            if(it == null) return false
            return getNearbyPlayer(it, blockPos) != null
        }
    }

    companion object {
        fun serverTick(level: Level, pos: BlockPos, state: BlockState, be: WorkbenchBE) {
            val player = getNearbyPlayer(level, pos)
            if (player != null && be.playerInRange == null) {
                be.triggerAnim("main", "rotating")
            } else if (player == null && be.playerInRange != null) {
                be.triggerAnim("main", "closed")
            }
            be.playerInRange = player
        }
    }
}

private fun getNearbyPlayer(level: Level, pos: BlockPos): Player? {
    return level.getNearestPlayer(pos.x.toDouble() + 0.5, pos.y.toDouble() + 0.5, pos.z.toDouble() + 0.5, 4.0, false)
}
